Fix backend ID size in BPF maps

brb opened the pull request:
For some reasons, the be323d8ca commit totally missed the fact that `loadbalancer.BackendID` is of the uin16 type and re-used the type for `backend_id` in the LB BPF maps. So, `backend_id` was uint16.

Meanwhile, the backend ID allocator allocated internally uint32, but the API call `AcquireBackendID` casteds IDs to `loadbalancer.BackendID`(=uint16). So, at least in theory (the cluster should have witnessed at least 2**16 different service endpoints), it could return the same IDs for different backends.

To avoid losing established connections, we need to migrate backend maps from `key(uint16)` to `key (uint32)` during bootstrap of cilium-agent. The removal of the old backend maps happens after all restored endpoints have been regenerated, so no interruptions in connections should be expected. 

**IMPORTANT** However, the downgrade will interrupt the connections. Also, probably the manual cleanup of CT maps is required to avoid the wrong endpoint routing problem, as most likely different backend IDs will be allocated;-/ Alternatively, we could maintain both maps...

Can be reviewed on per-commit basis.

Fixes: be323d8ca ("bpf,lbmap: Change backend ID to uint32")

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8041)
<!-- Reviewable:end -->

brb committed:
datapath: Fix backend_id type (uint32)

For some reasons, the be323d8ca commit totally missed the fact that
`loadbalancer.BackendID` is of the uin16 type and re-used the type
for `backend_id` in the LB BPF maps. So, `backend_id` was uint16.

Meanwhile, the backend ID allocator allocated internally uint32, but
the API call `AcquireBackendID` casteds IDs to
`loadbalancer.BackendID`(=uint16). So technically, it could returned
the same IDs for different backends.

Fixes: be323d8ca ("bpf,lbmap: Change backend ID to uint32")
Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
bpf: Add OpenIfExists and IsEmpty methods to bpf.Map

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Migrate existing backend maps

To avoid losing established connections, we need to migrate backend maps
from `key(uint16)` to `key (uint32)` during bootstrap of cilium-agent.

The removal of the old backend maps happens after all restored endpoints
have been regenerated, so no interruptions in connections should be expected.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm committed:
test bump image of upgrade / downgrade test to v1.5

Signed-off-by: André Martins <andre@cilium.io>

aanm commented on cilium/cmd/bpf_lb_list.go:
```diff
@@ -49,7 +48,7 @@ func dumpRevNat(serviceList map[string][]string) {
 	// It's safe to use the same map for both IPv4 and IPv6, as backend
 	// IDs are allocated from the same pool regardless the protocol
-	backendMap := make(map[loadbalancer.BackendID]lbmap.BackendValue)
+	backendMap := make(map[uint32]lbmap.BackendValue)
```
This change seems unnecessary.

brb replied:
Please see my comment below

aanm commented on pkg/maps/lbmap/types.go:
```diff
@@ -223,7 +223,7 @@ type Backend interface {
 
 	// Get backend identifier
-	GetID() loadbalancer.BackendID
+	GetID() uint32
```
Why are we changing this from `loadbalancer.BackendID` to `uint32` if the type `loadbalancer.BackendID` is set to `uint32`? (Same question applies to most of the changes in this commit)

brb replied:
The bug which this PR tries to fix was introduced because I decided to reuse the type definition. IMO, to prevent from such mishaps in the future, it's better to avoid representing BPF maps in Go with "higher-level" types.

aanm commented on test/helpers/cons.go:
```diff
@@ -174,7 +174,7 @@ const (
 	KubectlPolicyNameSpaceLabel = k8sConst.PolicyLabelNamespace
 
-	CiliumStableVersion      = "v1.4"
+	CiliumStableVersion      = "v1.5"
```
Maybe it's better if you cherry-pick the commit https://github.com/cilium/cilium/pull/8087/commits/196c182c938ff8c80e3bb577618bfd7ff469aac9

brb replied:
Good idea, going to cherry-pick after I've fixed that PR.

brb committed:
test: Do not set bpf-ct-global-tcp-max

As we are changing the stable version of Cilium in the upgrade test
to v1.5, we need to ensure that no CT TCP maps should be recreated.
Otherwise, the SVC migration tests get broken.

After this change, the size of the CT TCP maps is defined only in
"examples/kubernetes/1.*/cilium-cm.yaml".

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Disable legacy services for upgrades from >= v1.5

The current upgrade test does an upgrade from v1.5 to the latest
master.

Previously, in v1.5 manifests ("examples/kubernetes") we disabled
legacy services, but meanwhile in the test manifests we enabled them.
This led to removal of service v2 backends during the upgrade,
as no legacy services could have been found and thus, DeleteOrphanBackends
was called.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Add k8s test manifest files for Cilium v1.5

Otherwise, when running the upgrade-downgrade test, in the downgrade
step, the test executor will use the k8sT/manifests/cilium-ds-patch.yaml
(a fallback when one for the v1.5 is not found) which prevents from
changing the container image.

Signed-off-by: Martynas Pumputis <m@lambda.lt>
